django/applications/catmaid/static/js/widgets/connector-viewer.js

```diff
@@ -0,0 +1,1118 @@
+  var ConnectorViewer = function(skeletonModels)
+  {
+    this.widgetID = this.registerInstance();
+    this.idPrefix = `connector-viewer${this.widgetID}-`;
+
+    // This skeleton source takes care of internal skeleton management. It is
+    // not registered. It is the input skeleton sink, but the output is handled
+    // with a second source
+    var updateWithSkels = this.updateWithSkels.bind(this);
+    this.skeletonSource = new CATMAID.BasicSkeletonSource(this.getName() + " Input", {
+      register: false,
+      handleAddedModels: updateWithSkels,
+      handleChangedModels: updateWithSkels,
+      handleRemovedModels: updateWithSkels
+    });
+    // A skeleton source to collect results in
+    this.resultSkeletonSource = new CATMAID.BasicSkeletonSource(this.getName());
+
+    this.cache = new ConnectorViewerCache(this.skeletonSource);
+    this.currentConnectorOrder = [];
+    this.firstConnectorIdx = 0;
+
+    this.currentConnectorRelation = DEFAULT_CONNECTOR_RELATION;
+    this.dimensions = [DEFAULT_HEIGHT, DEFAULT_WIDTH];
+
+    this.sourceStackViewer = project.getStackViewers()[0];
+    this.stackViewers = [];
+    this.panelWindows = [];
+    this.showScaleBar = DEFAULT_SHOW_SCALE_BAR;
+
+    if (skeletonModels) {
+      this.skeletonSource.append(skeletonModels);
+    }
+  };
+
+  ConnectorViewer.prototype = {};
+  $.extend(ConnectorViewer.prototype, new InstanceRegistry());
+
+  ConnectorViewer.prototype.getName = function() {
+    return "Connector Viewer " + this.widgetID;
+  };
+
+  ConnectorViewer.prototype.destroy = function() {
+    this.closeStackViewers();
+    this.unregisterInstance();
+  };
+
+  /**
+   * Update the text describing which connectors are shown.
+   */
+  ConnectorViewer.prototype.updateShowingText = function() {
+    var total = this.currentConnectorOrder.length;
+    var start = Math.min(this.firstConnectorIdx + 1, total);
+    var stop = Math.min(this.firstConnectorIdx + this.dimensions[0] * this.dimensions[1], total);
+
+    var showingTextSelector = $(`#${this.idPrefix}showing`);
+    showingTextSelector.find(`.start`).text(start);
+    showingTextSelector.find(`.stop`).text(stop);
+    showingTextSelector.find(`.total`).text(total);
+  };
+
+  /**
+   *
+   * @param newPage zero-indexed
+   */
+  ConnectorViewer.prototype.changePage = function(newPage) {
+    var currentPageElement = document.getElementById(this.idPrefix + 'current-page');
+
+    var total = this.currentConnectorOrder.length;
+
+    if (total === 0) {
+      currentPageElement.value = 1;
+      this.update();
+      return 0;
+    }
+
+    var newFirstConnectorIdx = newPage * this.dimensions[0] * this.dimensions[1];
+
+    if (this.firstConnectorIdx === newFirstConnectorIdx) {  // page may not be changing
+      this.update();
+      return Number(currentPageElement.value) - 1;
+    } else if (newPage < 0 || newFirstConnectorIdx >= total) {  // page out of bounds
+      alert('This page does not exist! Returning to page 1.');
```

Comment:
Instead of `alert()`, I think `CATMAID.error()` would be nicer, because most errors are reported that way. However, maybe `CATMAID.warn()` would be enough to not interrupt too much due a wrong page number.